packages/datadog-instrumentations/src/jest.js

```diff
@@ -1277,77 +1279,81 @@ addHook({
-    return
-  }
-  if (code === JEST_WORKER_COVERAGE_PAYLOAD_CODE) { // datadog coverage payload
-    workerReportCoverageCh.publish(data)
-    return
-  }
-  if (code === JEST_WORKER_LOGS_PAYLOAD_CODE) { // datadog logs payload
-    workerReportLogsCh.publish(data)
-    return
+function sendWrapper (send) {
+  return function (request) {
+    if (!isKnownTestsEnabled && !isTestManagementTestsEnabled && !isImpactedTestsEnabled) {
+      return send.apply(this, arguments)
+    }
+    const [type] = request
+
+    // https://github.com/jestjs/jest/blob/1d682f21c7a35da4d3ab3a1436a357b980ebd0fa/packages/jest-worker/src/workers/ChildProcessWorker.ts#L424
+    if (type === CHILD_MESSAGE_CALL) {
+      // This is the message that the main process sends to the worker to run a test suite (=test file).
+      // In here we modify the config.testEnvironmentOptions to include the known tests for the suite.
+      // This way the suite only knows about the tests that are part of it.
+      const args = request.at(-1)
+      if (args.length > 1) {
+        return send.apply(this, arguments)
+      }
+      if (!args[0]?.config) {
+        return send.apply(this, arguments)
+      }
+      const [{ globalConfig, config, path: testSuiteAbsolutePath }] = args
+      const testSuite = getTestSuitePath(testSuiteAbsolutePath, globalConfig.rootDir || process.cwd())
+      const suiteKnownTests = knownTests?.jest?.[testSuite] || []
+
+      const suiteTestManagementTests = testManagementTests?.jest?.suites?.[testSuite]?.tests || {}
+
+      const suiteModifiedTests = Object.keys(modifiedTests).length > 0
+        ? modifiedTests
+        : {}
+
+      args[0].config = {
+        ...config,
+        testEnvironmentOptions: {
+          ...config.testEnvironmentOptions,
+          _ddKnownTests: suiteKnownTests,
+          _ddTestManagementTests: suiteTestManagementTests,
+          _ddModifiedTests: suiteModifiedTests
+        }
+      }
+    }
+    return send.apply(this, arguments)
   }
-  return _onMessage.apply(this, arguments)
 }
 
-const enqueueWrapper = enqueue => function (...args) {
-  const originalOnStart = args[0].onStart
-  args[0].onStart = (worker) => {
-    if (worker && !worker._ddWorker) {
-      worker._child.send = sendWrapper(worker._child.send)
-      worker._onMessage = onMessageWrapper(worker._onMessage)
-      worker._child.on('message', worker._onMessage.bind(worker))
-      worker._ddWorker = true
-    }
-    originalOnStart(worker)
+function enqueueWrapper (enqueue) {
+  return function () {
+    shimmer.wrap(arguments[0], 'onStart', onStart => function (worker) {
+      if (worker && !wrappedWorkers.has(worker)) {
+        shimmer.wrap(worker._child, 'send', sendWrapper)
+        shimmer.wrap(worker, '_onMessage', onMessageWrapper)
+        worker._child.on('message', worker._onMessage.bind(worker))
+        wrappedWorkers.add(worker)
+      }
+      return onStart.apply(this, arguments)
+    })
+    return enqueue.apply(this, arguments)
```

Comment:
changes:
* `function enqueueWrapper` instead of `const enqueueWrapper` (just more consistent and it has other advantages)
* use `arguments` instead of `...args`
* use `wrappedWorkers` instead of `_ddWorker`
* use `shimmer.wrap` whenever possible instead of directly wrapping